### PR TITLE
Dont Re-export sp-arithmetic from sp-runtime

### LINKED ISSUE #12040: Don't re-export `sp-arithmetic` from `sp-runtime`
Currently, we are re-exporting this crate from sp-runtime. This makes it be a part of a lot of other crates, since sp-runtime is itself a wildly used crate. Consequently, this has prevented us from adding other 3rd party dependencies to sp-arithmetic, since it would naturally leak into the wider sp-runtime as well (https://github.com/encointer/substrate-fixed).

because of this, and the spirit of dependency isolation, we should explicitly import sp-arithmetic when needed, and remove the re-export from sp-runtime. 

This will be a big PR, but should be very easy to do.

### PR DESCRIPTION
**Pull request by MrishoLukamba**
Fixes #12040
Polkadot address: 133WyzJ4cEEbQ4twhT3nQpoXYeTCzeKn8kWfRfEi4XDF7wKh
The Pr is complete.
The changes will be made as small as possible untill all dependencies from arithmetic re-exported from sp runtime is not re-exported anymore.

### CODE REVIEW

**Comment by MrishoLukamba** [comment]
All done, Please review

File: bin/node-template/node/src/benchmarking.rs [hunk 1]
```diff
@@ -29,8 +29,8 @@ use sc_client_api::BlockBackend;
 use sp_core::{Encode, Pair};
 use sp_inherents::{InherentData, InherentDataProvider};
 use sp_keyring::Sr25519Keyring;
-use sp_runtime::{OpaqueExtrinsic, SaturatedConversion};
-
+use sp_runtime::{OpaqueExtrinsic};
+use sp_arithmetic::traits::SaturatedConversion;
```
Comment (bkchr) [inline bin/node-template/node/src/benchmarking.rs:33]:
```suggestion
use sp_runtime::OpaqueExtrinsic;
use sp_arithmetic::traits::SaturatedConversion;

```

File: bin/node-template/runtime/Cargo.toml [hunk 2]
```diff
@@ -38,7 +38,7 @@ sp-session = { version = "4.0.0-dev", default-features = false, path = "../../..
 sp-std = { version = "4.0.0", default-features = false, path = "../../../primitives/std" }
 sp-transaction-pool = { version = "4.0.0-dev", default-features = false, path = "../../../primitives/transaction-pool" }
 sp-version = { version = "5.0.0", default-features = false, path = "../../../primitives/version" }
-
+sp-arithmetic = { version = "5.0.0", default-features = false, path = "../../../primitives/arithmetic" }
```
Comment (bkchr) [inline bin/node-template/runtime/Cargo.toml:41]:
```suggestion
sp-arithmetic = { version = "5.0.0", default-features = false, path = "../../../primitives/arithmetic" }

```

File: bin/node/cli/src/service.rs [hunk 3]
```diff
@@ -35,9 +35,9 @@ use sc_service::{config::Configuration, error::Error as ServiceError, RpcHandler
 use sc_telemetry::{Telemetry, TelemetryWorker};
 use sp_api::ProvideRuntimeApi;
 use sp_core::crypto::Pair;
-use sp_runtime::{generic, traits::Block as BlockT, SaturatedConversion};
+use sp_runtime::{generic, traits::Block as BlockT};
 use std::sync::Arc;
-
+use sp_arithmetic::traits::SaturatedConversion;
```
Comment (bkchr) [inline bin/node/cli/src/service.rs:40]:
```suggestion
use sp_arithmetic::traits::SaturatedConversion;

```

File: bin/node/primitives/Cargo.toml [hunk 4]
```diff
@@ -20,7 +20,7 @@ frame-system = { version = "4.0.0-dev", default-features = false, path = "../../
 sp-application-crypto = { version = "6.0.0", default-features = false, path = "../../../primitives/application-crypto" }
 sp-core = { version = "6.0.0", default-features = false, path = "../../../primitives/core" }
 sp-runtime = { version = "6.0.0", default-features = false, path = "../../../primitives/runtime" }
-
+sp-arithmetic = { version = "5.0.0", default-features = false, path = "../../../primitives/arithmetic" }
```
Comment (bkchr) [inline bin/node/primitives/Cargo.toml:23]:
```suggestion
sp-arithmetic = { version = "5.0.0", default-features = false, path = "../../../primitives/arithmetic" }

```

**Comment by MrishoLukamba** [comment]
> Next time please put some more effort into this. Especially when you ping a lot of people.

Okey

**Comment by bkchr** [comment]
@MrishoLukamba the CI is still failing.

**Comment by MrishoLukamba** [comment]
> @MrishoLukamba the CI is still failing.

What can be the reason because it is compiling

**Comment by ggwpez** [comment]
@MrishoLukamba you can click on the failing CI jobs and see why.  
Eg https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1960850:  
```pre
error[E0432]: unresolved imports `sp_runtime::Perbill`, `sp_runtime::Percent`
   --> primitives/npos-elections/src/phragmms.rs:235:19
    |
235 |     use sp_runtime::{Perbill, Percent};
    |                      ^^^^^^^  ^^^^^^^ no `Percent` in the root
    |                      |
    |                      no `Perbill` in the root
```

**Comment by shawntabrizi** [comment]
you also need to call `cargo +nightly fmt`

**Comment by MrishoLukamba** [comment]
> you also need to call `cargo +nightly fmt`

Updating the docs also

**Comment by MrishoLukamba** [comment]
> [](/paritytech/substrate/actions/runs/3289213265/jobs/5420475860)
> [ @github-actions ](/apps/github-actions)
> ** Assign reviewers / pr-custom-review (pull_request) ** Successful in 11s
> Required

Now the CI complains but when I check the job its okey and everything is fine.

**Comment by ggwpez** [comment]
> Now the CI complains but when I check the job its okey and everything is fine.

CI is still red. You click on "Details" and then on the pipeline and on its failing job, eg https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1971086

**Comment by MrishoLukamba** [comment]
the CI which are failing now its not because of the codebase anymore. So its done

**Comment by shawntabrizi** [comment]
@MrishoLukamba unfortunately... you now need to repeat this process for Polkadot and Cumulus for us to merge this PR

See: https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.adoc#updating-polkadot-as-well

**Comment by MrishoLukamba** [comment]
Okey

**Review by acatangiu — APPROVED** [review]

File: bin/node/executor/tests/fees.rs [hunk 5]
```diff
@@ -28,7 +28,8 @@ use kitchensink_runtime::{
 };
 use node_primitives::Balance;
 use node_testing::keyring::*;
-use sp_runtime::{traits::One, Perbill};
+use sp_arithmetic::Perbill;
+use sp_runtime::traits::One;
```
Comment (kianenigma) [inline bin/node/executor/tests/fees.rs:32]: Shouldn't this trait also be in arithmetic? `One` and `Zero`.

File: bin/node/testing/src/client.rs [hunk 6]
```diff
@@ -1,4 +1,5 @@
 // This file is part of Substrate.
+// This file is part of Substrate.
```
Comment (kianenigma) [inline bin/node/testing/src/client.rs:2]:
```suggestion
```

File: frame/election-provider-multi-phase/src/benchmarking.rs [hunk 7]
```diff
@@ -27,8 +27,7 @@ use frame_support::{
 };
 use frame_system::RawOrigin;
 use rand::{prelude::SliceRandom, rngs::SmallRng, SeedableRng};
-use sp_arithmetic::{per_things::Percent, traits::One};
-use sp_runtime::InnerOf;
+use sp_arithmetic::{per_things::Percent, traits::One, InnerOf};
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/benchmarking.rs:30]:
```suggestion
use sp_arithmetic::{Percent, traits::One, InnerOf};
```

All types should be re-exported from the root crate, all traits must be in `traits`.

File: frame/nomination-pools/src/lib.rs [hunk 8]
```diff
@@ -280,13 +280,10 @@ use frame_support::{
...
+use sp_arithmetic::{FixedPointNumber, FixedPointOperand};
 use sp_core::U256;
-use sp_runtime::{
-	traits::{
-		AccountIdConversion, Bounded, CheckedAdd, CheckedSub, Convert, Saturating, StaticLookup,
-		Zero,
-	},
-	FixedPointNumber, FixedPointOperand,
+use sp_runtime::traits::{
+	AccountIdConversion, Bounded, CheckedAdd, CheckedSub, Convert, Saturating, StaticLookup, Zero,
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:286]: here as well.

File: frame/offences/benchmarking/src/lib.rs [hunk 9]
```diff
@@ -25,13 +25,12 @@ mod mock;
...
+use frame_election_provider_support::sp_arithmetic;
 use frame_support::traits::{Currency, Get, ValidatorSet, ValidatorSetWithIdentification};
 use frame_system::{Config as SystemConfig, Pallet as System, RawOrigin};
 
-use sp_runtime::{
-	traits::{Convert, Saturating, StaticLookup, UniqueSaturatedInto},
-	Perbill,
-};
+use sp_arithmetic::Perbill;
+use sp_runtime::traits::{Convert, Saturating, StaticLookup, UniqueSaturatedInto};
```
Comment (kianenigma) [inline frame/offences/benchmarking/src/lib.rs:33]: And here.

**Review by kianenigma — CHANGES_REQUESTED** [review]
I see some re-exports of arithmetic stiff left from `sp_runtime`. Do they have a reason that I miss?

File: bin/node/executor/tests/fees.rs (hunk 5 above)
Comment (MrishoLukamba) [reply bin/node/executor/tests/fees.rs:32]: One and Zero are in sp_runtime

Comment (MrishoLukamba) [reply bin/node/executor/tests/fees.rs:32]: and you are looking at earlier commits

**Comment by bkchr** [comment]
Ping @MrishoLukamba

**Comment by juangirini** [comment]
@MrishoLukamba FYI #12040 has been closed so your PR is not going to be merged, though we appreciate your contribution. Thanks!

**Pull request closed without merging**